### ISSUE TITLE
Proj0243 false positive: <GenerateSBOM> task element picked up instead of the property

### ISSUE THREAD

**Issue by JohnVerheij** [dotnet-project-file-analyzers/dotnet-project-file-analyzers#800]
When `<GenerateSBOM>true</GenerateSBOM>` is set in `Directory.Build.props` and a `<GenerateSBOM>` MSBuild task is invoked inside a `<Target>` in `Directory.Build.targets`, Proj0243 fires on the task element.

## Repro

`global.json`:

```json
{
  "sdk": {
    "version": "10.0.300",
    "rollForward": "latestFeature"
  }
}
```

`Directory.Packages.props`:

```xml
<Project>
  <PropertyGroup>
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
  </PropertyGroup>
  <ItemGroup>
    <PackageVersion Include="Microsoft.Sbom.Targets" Version="4.1.5" />
  </ItemGroup>
  <ItemGroup>
    <GlobalPackageReference Include="DotNetProjectFile.Analyzers" Version="1.14.0" PrivateAssets="all" />
  </ItemGroup>
</Project>
```

`Directory.Build.props`:

```xml
<Project>
  <PropertyGroup>
    <TargetFramework>net10.0</TargetFramework>
    <Nullable>enable</Nullable>
    <GenerateSBOM>true</GenerateSBOM>
  </PropertyGroup>
</Project>
```

`Directory.Build.targets`:

```xml
<Project>
  <Target Name="GenerateSbomOnPublish" AfterTargets="Publish"
          Condition="'$(GenerateSBOM)' == 'true' AND '$(IsPackable)' != 'true'">
    <GenerateSBOM
      BuildDropPath="$(PublishDir)"
      PackageName="$(MSBuildProjectName)"
      PackageVersion="$(Version)"
      PackageSupplier="Example"
      NamespaceBaseUri="https://example.com/sbom" />
  </Target>
</Project>
```

`MyLib/MyLib.csproj`:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <IsPackable>true</IsPackable>
    <Version>1.0.0</Version>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Sbom.Targets" PrivateAssets="all" />
  </ItemGroup>
</Project>
```

Build with `dotnet build MyLib/MyLib.csproj` produces:

```
Directory.Build.targets(4,5): warning Proj0243: Enable SBOM generation with <GenerateSBOM> is 'true' or define the <IsPackable> node with value 'false'
```

`(4,5)` is the position of the `<GenerateSBOM` task element. Removing `Directory.Build.targets` from the repo silences the warning.

## Root cause

`MsBuildProject.Property<TNode>()` returns the first matching node from `SelfAndDirectoryProps()`, which yields `Directory.Build.targets` before `Directory.Build.props`. `<GenerateSBOM>` parses to the same `GenerateSBOM : Node<bool?>` type whether the parent is `<PropertyGroup>` or `<Target>`; the task element has no inner text so its `Value` is not `true`. `GenerateSbom.Register` then evaluates `property.Value is not true` on the task element and reports the diagnostic at its location.

Relevant files:

- `src/DotNetProjectFile.Analyzers/MsBuild/MsBuildProject.Property.cs`
- `src/DotNetProjectFile.Analyzers/MsBuild/MsBuildProject.cs`
- `src/DotNetProjectFile.Analyzers/MsBuild/GenerateSBOM.cs`
- `src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/GenerateSbom.cs`

**Comment by Corniel** [comment]
So, in other words, It only checks for `<PropertyItems>`, and not nodes within `<Target>`? Seems like something we should take into account indeed.

**Comment by CptWesley** [comment]
@Corniel I think it's the other way around. It's finding the one in `<Target>`, which is not the one we should be searching for.

**Comment by JohnVerheij** [comment]
The analyzer is finding the `<GenerateSBOM>` element inside `<Target>`, not missing it. `MsBuildProject.Property<TNode>()` walks `SelfAndDirectoryProps()` in order (`Directory.Build.targets` first, then `Directory.Build.props`), and `<GenerateSBOM>` parses to the same `Node<bool?>` type whether its parent is `<PropertyGroup>` or `<Target>`, so the task element wins the lookup. Its `Value` is then evaluated as `not true` (task elements have no inner text) and the diagnostic fires at its location. The property in props itself is correct; only the task element in targets is wrongly flagged.

**Comment by Corniel** [comment]
@JohnVerheij Can you review the linked PR?
